web/packages/teleterm/src/ui/Search/pickers/ActionPicker.tsx

```diff
@@ -281,30 +307,71 @@ const ExtraTopComponents = (props: {
   }
 };
 
+/**
+ * ActionPickerStatus helps with displaying ExtraTopComponents. It has two goals:
+ *
+ *   * Encapsulate business logic so that anything that ExtraTopComponents renders can just read
+ *     ActionPickerStatus fields.
+ *   * Represent only valid UI states. For example, inputState 'no-input' doesn't have hasNoResults
+ *     field as this field would make no sense in a situation where no search requests were made.
+ *
+ * As you may notice, ActionPickerStatus doesn't say whether the search request is in progress or
+ * not, simply because displaying the progress bar is handled by another component. The questions
+ * answered by ActionPickerStatus are valid to ask no matter what the state of the request is.
```

Comment:
(philosophical tangent ahead)

To elaborate on that, for example `ExtraTopComponents` only needs to know if there are no search results, no matter what the status of the request is.

If the request is in progress, the `hasNoResults` is false because we didn't finish fetching yet.

If the request is done, `hasNoResults` can be true/false depending on what the response is.

Had we returned request status next to `hasNoResults`, it'd make it possible to introduce nonsensical states, e.g. `hasNoResults: true, requestState: 'processing'`. But since we don't tell what the status of the request is, every combination of fields that are currently present in this type makes sense.

---

This is all possible only because `getActionPickerStatus` knows what the actual status of the request is, so `ActionPickerStatus` can act as a proxy when answering the questions, without the components in `ExtraTopComponents` having to worry whether the request is in progress or not. This is not the equivalent of [forgetting to check the `isLoading` flag](http://blog.jenkster.com/2016/06/how-elm-slays-a-ui-antipattern.html).